Ficr prog sample

bama-nordic opened the pull request:
This PR adds shell functions to program FICR (Factory Information Configuration Registers) residing in the OTP memory of nrRF7002 device.

NordicBuilder commented:
You can find the documentation preview for this PR at this <a href='http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-8952
'>link</a>. It will be updated about 10 minutes after the documentation build succeeds.

*Note: This comment is automatically posted by the Documentation Publishing GitHub Action.*

NordicBuilder commented:
### Integration test specification

| Test Module | File based changes | Manually selected | West overwrite |
| ----------- | :----------------: | :---------------: | :------------: |
| test-sdk-wifi | X | | |


[Detailed information of selected test modules](https://jenkins-ncs.nordicsemi.no/job/latest/job/sub/job/test-fw-nrfconnect-nrf/job/master/27518/artifact/test-spec/test-spec-summary.yml/*view*/)

*Note: This message is automatically posted and updated by the CI*

krish2718 commented on drivers/wifi/nrf700x/zephyr/src/qspi/inc/ficr_prog.h:
```diff
@@ -0,0 +1,86 @@
+ */
+
+/**
+ * @brief Header containing common functions for RPU hardware interaction
+ * using QSPI and SPI that can be invoked by shell or the driver.
```
Need to modify the text, this is from rpu_fw_if.c

bama-nordic replied:
done

krish2718 commented on drivers/wifi/nrf700x/zephyr/src/qspi/inc/ficr_prog.h:
```diff
@@ -0,0 +1,86 @@
+#include <stdio.h>
+#include <stdlib.h>
+
+//REGION PROTECT : OTP Address offsets (word offsets)
```
/* */ is preferred for oneline comments.

bama-nordic replied:
done

krish2718 commented on samples/wifi/radio_test/src/nrf_wifi_radio_ficr_shell.c:
```diff
@@ -0,0 +1,335 @@
+	uint32_t ret, err;
+	uint32_t val[OTP_MAX_WORD_LEN];
+
+    // read all the OTP memory
```
wrong indent

bama-nordic replied:
Done

krish2718 reviewed (changes requested):
LGTM, just some nits.

richabp reviewed (commented):
Few general comments:
1. Are Radio test sample and FICR sample parts of the same sample? If yes, then we need to add some relevant content at the main page Wi-Fi: Test explaining this.

2. Are Radio test sample and FICR sample two different samples? If yes, can we keep both the samples separate e.g Radio Test and Radio FICR sample? In this case, we will not need any content at the Wi-Fi: Test page and we can remove this page.

3. RST files have one sentence per line so need to change wherever it is applicable.

bama-nordic commented:
> Few general comments:
> 
> 1. Are Radio test sample and FICR sample parts of the same sample? If yes, then we need to add some relevant content at the main page Wi-Fi: Test explaining this.
> 2. Are Radio test sample and FICR sample two different samples? If yes, can we keep both the samples separate e.g Radio Test and Radio FICR sample? In this case, we will not need any content at the Wi-Fi: Test page and we can remove this page.
> 3. RST files have one sentence per line so need to change wherever it is applicable.

Thanks Richa. They are both parts of the Wi-Fi Radio Test sample. I have added this information in the main page as below - pls review.
"This sample supports the Radio Test sub-commands and sub-commands to program
the FICR fields defined in the nRF7002 OTP memory"

sachinthegreen commented on drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c:
```diff
@@ -0,0 +1,452 @@
+	int otp_mem_status = 0;
+	int poll = 0;
+
+	while ((otp_mem_status != OTP_READ_VALID) && (poll < 100)) {
```
Should this also be checked against (otp_mem_status & 0x2) as is being done below?

bama-nordic replied:
Done

sachinthegreen commented on drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c:
```diff
@@ -0,0 +1,452 @@
+	int otp_mem_status = 0;
+	int poll = 0;
+
+	while ((otp_mem_status != OTP_WR_DONE) && (poll < 100)) {
```
Should this also be checked against (otp_mem_status & 0x1) as is being done below?

bama-nordic replied:
Done

sachinthegreen commented on drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c:
```diff
@@ -0,0 +1,452 @@
+{
+	uint32_t val = OTP_READ_MODE;
+	int32_t err;
+
```
No need of these variables, since the values can be directly used below.

bama-nordic replied:
Done

sachinthegreen commented on drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c:
```diff
@@ -0,0 +1,452 @@
+{
+	uint32_t val = OTP_BYTE_WRITE_MODE;
+	int32_t err;
+
```
No need of these variables, since the values can be directly used below.

bama-nordic replied:
Done

sachinthegreen commented on drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c:
```diff
@@ -0,0 +1,452 @@
+		return err;
+	}
+	return 0;
+
```
We can just do return poll_otp_ready(). instead of the above statements.

bama-nordic replied:
Done

sachinthegreen commented on drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c:
```diff
@@ -0,0 +1,452 @@
+	}
+	read_word(rd_addr, read_val);
+	return *read_val;
+
```
If we are going to return the read_val anyway, it might not be necessary to get uint32_t *read_val as a parameter.

bama-nordic replied:
Done

sachinthegreen commented on drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c:
```diff
@@ -0,0 +1,452 @@
+static int  write_otp_location(uint32_t otp_location_offset, uint32_t otp_data)
+{
+	int ret;
+
```
ret does not seem to be necessary.

bama-nordic replied:
Done

sachinthegreen commented on drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c:
```diff
@@ -0,0 +1,452 @@
+static int otp_rd_voltage_1V8(void)
+{
+	uint32_t val = OTP_VOLTCTRL_1V8;
+	int32_t err;
```
val does not seem to be necessary.

bama-nordic replied:
Done

sachinthegreen commented on drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c:
```diff
@@ -0,0 +1,452 @@
+{
+	int32_t		err;
+	int32_t		mask_val;
+
```
For consistency, keep only 1 space between data type and variable name.

bama-nordic replied:
Done

sachinthegreen commented on drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c:
```diff
@@ -0,0 +1,452 @@
+	switch (otp_addr) {
+
+	case REGION_PROTECT:
+		write_otp_location(REGION_PROTECT, write_val[0]);
```
return value of write_otp_location not being considered? (Generic comment for rest of cases too).

bama-nordic replied:
error condition is logged in the function, so this is fine.

sachinthegreen commented on drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c:
```diff
@@ -0,0 +1,452 @@
+		break;
+
+	case QSPI_KEY:
+		mask_val = QSPI_KEY_FLAG_MASK;
```
Should this be reading the pre-existing mask value and then Oring it with QSPI_KEY_FLAG_MASK?  (Generic comment for all other cases below too)

bama-nordic replied:
Not necessary since we dont require s read-modify-write operation here.

sachinthegreen commented on samples/wifi/radio_test/src/nrf_wifi_radio_ficr_shell.c:
```diff
@@ -0,0 +1,346 @@
+	uint32_t val[OTP_MAX_WORD_LEN];
+
+	/* read all the OTP memory */
+	err = read_otp_memory(0, &val[0], OTP_MAX_WORD_LEN);
```
This could be made as `err = read_otp_memory(0, val, sizeof(val));` so that it will work even if size of val changes in the future.

bama-nordic replied:
will look into this for the next revision

sachinthegreen commented on samples/wifi/radio_test/src/nrf_wifi_radio_ficr_shell.c:
```diff
@@ -0,0 +1,346 @@
+	uint32_t ret, err;
+
+
+	err = read_otp_memory(0, &val[0], OTP_MAX_WORD_LEN);
```
This could be made as `err = read_otp_memory(0, val, sizeof(val));` so that it will work even if size of val changes in the future.

bama-nordic replied:
Will look into this for the next revision

bama-nordic commented on drivers/wifi/nrf700x/zephyr/src/qspi/inc/ficr_prog.h:
```diff
@@ -0,0 +1,86 @@
+#define PRODTEST_TRIM_LEN       15
+
+#define ENABLE_OTP_WRITE        1
+
```
making this os agnostic is something we'll take up in the next revision. AI on me to raise a Jira issue on this.

bama-nordic commented on drivers/wifi/nrf700x/zephyr/src/qspi/inc/ficr_prog.h:
```diff
@@ -0,0 +1,86 @@
+int read_otp_memory(uint32_t otp_addr, uint32_t *read_val, int len);
+uint32_t check_protection(uint32_t *buff, uint32_t off1, uint32_t off2,
+		    uint32_t off3, uint32_t off4);
+
```
making this os agnostic is something we'll take up in the next revision. AI on me to raise a Jira issue on this.

bama-nordic replied:
Better to keep this function generic for future use

bama-nordic commented on drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c:
```diff
@@ -0,0 +1,452 @@
+
+LOG_MODULE_DECLARE(otp_prog, CONFIG_WIFI_LOG_LEVEL);
+
+static void write_word(uint32_t addr, uint32_t data)
```
making this os agnostic is something we'll take up in the next revision. AI on me to raise a Jira issue on this.

bama-nordic commented on drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c:
```diff
@@ -0,0 +1,452 @@
+}
+
+static void read_word(uint32_t addr, uint32_t *data)
+{
```
making this os agnostic is something we'll take up in the next revision. AI on me to raise a Jira issue on this.

bama-nordic commented on drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c:
```diff
@@ -0,0 +1,452 @@
+/*
```
making this os agnostic is something we'll take up in the next revision. AI on me to raise a Jira issue on this.

bama-nordic commented on drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c:
```diff
@@ -0,0 +1,452 @@
+static int otp_wr_voltage_2V5(void)
+{
+	uint32_t val = OTP_VOLTCTRL_2V5;
+	int32_t err;
```
Done

bama-nordic commented on drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c:
```diff
@@ -0,0 +1,452 @@
+
+static int req_otp_standby_mode(void)
+{
+
```
Done

bama-nordic commented on drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c:
```diff
@@ -0,0 +1,452 @@
+
+	uint32_t val;
+	int32_t otp_ready_exception;
+
```
Done

bama-nordic commented on drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c:
```diff
@@ -0,0 +1,452 @@
+
+uint32_t check_protection(uint32_t *buff, uint32_t off1, uint32_t off2,
+		    uint32_t off3, uint32_t off4)
+{
```
leaving this for future use

bama-nordic commented on drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c:
```diff
@@ -0,0 +1,452 @@
+static void set_otp_timing_reg_40mhz(void)
+{
+	uint32_t val;
+
```
Done

richabp commented on samples/wifi/radio_test/ficr.rst:
```diff
@@ -0,0 +1,303 @@
+   :local:
+   :depth: 2
+
+The Wi-Fi Radio Factory Information Configuration Registers (FICR) programming
+sample demonstrates how to program the user region of FICR parameters on the
+development kit using a set of predefined commands. Please note that all the
+FICR registers are stored in a One Time Programmable (OTP) memory and
+consequently, write commands are DESTRUCTIVE, and once written, the contents
+of OTP registers cannot be re-programmed.
```
Moved the second sentence to the Overview section as a Note.
```suggestion
The Wi-Fi Radio Factory Information Configuration Registers (FICR) programming sample demonstrates how to program the user region of FICR parameters on the development kit. It uses a set of predefined commands.
```

richabp commented on samples/wifi/radio_test/ficr.rst:
```diff
@@ -0,0 +1,303 @@
+Building and running
+********************
+
+.. |sample path| replace:: :file:`samples/wifi/radio_test`
```
```suggestion
.. |sample path| replace:: :file:`samples/wifi/ficr`
```

bama-nordic replied:
OK done.

bama-nordic commented on samples/wifi/radio_test/ficr.rst:
```diff
@@ -0,0 +1,303 @@
+FICR registers are stored in a One Time Programmable (OTP) memory and
+consequently, write commands are DESTRUCTIVE, and once written, the contents
+OTP registers cannot be re-programmed.
+
+Please read the FICR section of the product specifications carefully before
+using using these commands.
```
Yes, it is very important that the person using these functions refers to the register definitions in the product specs before attempting OTP programming

richabp replied:
Ok, then I believe we can add this line in the Note (already there in the Overview section) as well.

bama-nordic commented on samples/wifi/radio_test/README.rst:
```diff
@@ -1,372 +1,16 @@
-   :local:
-   :depth: 2
+This sample supports the Radio Test sub-commands and sub-commands to program
+the FICR fields defined in the nRF7002 OTP memory
```
OK done.

bama-nordic commented on samples/wifi/radio_test/ficr.rst:
```diff
@@ -0,0 +1,303 @@
+To run the commands, connect to the development kit through the serial port and send shell commands.
+Zephyr's :ref:`zephyr:shell_api` module is used to handle the commands.
+
+You can start running ``wifi_radio_ficr_prog`` subcommands to read or write OTP registers
```
OK done.

bama-nordic commented on samples/wifi/radio_test/ficr.rst:
```diff
@@ -0,0 +1,303 @@
+     - N/A
+     - N/A
+     - N/A
+     - reads out the OTP status of user region and status of each field (programmed or not)
```
OK done.

bama-nordic commented on samples/wifi/radio_test/ficr.rst:
```diff
@@ -0,0 +1,303 @@
+     - N/A
+     - N/A
+     - N/A
+     - reads out all the OTP parameters (excluding QSPI_KEY which cannot be read)
```
OK done.

bama-nordic commented on samples/wifi/radio_test/ficr.rst:
```diff
@@ -0,0 +1,303 @@
+     - QSPI_KEY
+     - 0x110
+     - arg1 arg2 arg3 arg4
+     - All 4 key arguments are 32-bit values forming the required 128-bit (Q)SPI key for encryption.
```
OK done.

bama-nordic commented on samples/wifi/radio_test/ficr.rst:
```diff
@@ -0,0 +1,303 @@
+     - MAC_ADDRESS1
+     - 0x128
+     - arg1 arg2
+     - If MAC addr is AA:BB:CC:DD:EE:FF, then arg1=0xDDCCBBAA and arg2=0xFFEE
```
OK done.

bama-nordic commented on samples/wifi/radio_test/ficr.rst:
```diff
@@ -0,0 +1,303 @@
+     - CALIB_MAXPOW2G4
+     - 0x13c
+     - arg
+     - arg is 32-bit value specifying Max output power for DSSS, MCS0, MCS7 (DSSS in bits[7:0] - unsigned). Measured in channel 7. Bits [31:24] unused.
```
OK done.

bama-nordic commented on samples/wifi/radio_test/ficr.rst:
```diff
@@ -0,0 +1,303 @@
+     - CALIB_MAXPOW5G0MCS7
+     - 0x140
+     - arg
+     - arg is 32-bit value specifying Max output power for MCS7 in channel 36, 100, 165 (channel 36 in bits[7:0] - unsigned). Bits [31:24] unused.
```
OK done.

bama-nordic commented on samples/wifi/radio_test/ficr.rst:
```diff
@@ -0,0 +1,303 @@
+     - CALIB_MAXPOW5G0MCS0
+     - 0x144
+     - arg
+     - arg is 32-bit value specifying Max output power for MCS0 in channel 36, 100, 165 (channel 36 in bits[7:0] - unsigned). Bits [31:24] unused.
```
OK done.

bama-nordic commented on samples/wifi/radio_test/ficr.rst:
```diff
@@ -0,0 +1,303 @@
+     - 0x154
+     - arg
+     - | Register default enable control.
+       | arg is 32-bit value where a specific bit set to '0' implies the corresponding OTP field is programmed. The bit & field mapping are
```
OK done.

bama-nordic commented on samples/wifi/radio_test/ficr.rst:
```diff
@@ -0,0 +1,303 @@
+     - CALIB_XO
+     - 0x130
+     - arg
+     - arg is 7-bit unsigned XO value. Bits [31:7] unused. Adjusts capacitor bank, 0 : Lowest capacitance (Highest frequency), 255 : Highest capacitance (Lowest frequency)
```
OK done.

bama-nordic commented on samples/wifi/radio_test/ficr.rst:
```diff
@@ -0,0 +1,303 @@
+     - CALIB_PDADJMCS7
+     - 0x134
+     - arg
+     - arg is 32-bit power detector adjustment for MCS7 for channel 7, 36, 100, 165 (channel 7 in bits[7:0] - signed)
```
OK done.

richabp replied:
I guess it's missed.

bama-nordic commented on samples/wifi/radio_test/ficr.rst:
```diff
@@ -0,0 +1,303 @@
+     - CALIB_PDADJMCS0
+     - 0x138
+     - arg
+     - arg is 32-bit power detector adjustment for MCS0 for channel 7, 36, 100, 165 (channel 7 in bits[7:0] - signed)
```
OK done.

richabp replied:
I guess it's missed.

bama-nordic commented on samples/wifi/radio_test/ficr.rst:
```diff
@@ -0,0 +1,303 @@
+
+#. |connect_kit|
+#. |connect_terminal|
+#. Reference command interface
```
OK done.

bama-nordic commented on samples/wifi/radio_test/ficr.rst:
```diff
@@ -0,0 +1,303 @@
+
+          wifi_radio_ficr_prog <subcommand> [Offset] [arg1] [arg2] .. [argN]
+
+#. Example commands:
```
OK done.

bama-nordic commented on samples/wifi/radio_test/ficr.rst:
```diff
@@ -0,0 +1,303 @@
+
+   #. To read the status of OTP Region
+
+      Execute the following command:
```
OK done.

bama-nordic commented on samples/wifi/radio_test/ficr.rst:
```diff
@@ -0,0 +1,303 @@
+
+   #. To write different locations of the OTP, follow the below instructions in conjunction with the tables describing
+
+      Execute the following sequence of commands to program MAC0 address to F0:CE:36:00:00:4A:
```
OK done.

richabp commented on samples/wifi/radio_test/ficr.rst:
```diff
@@ -0,0 +1,304 @@
+
+#. |connect_kit|
+#. |connect_terminal|
+#. Use the following reference command interface to read or write the OTP params:
+
+      .. code-block:: console
+
+          wifi_radio_ficr_prog <subcommand> [Offset] [arg1] [arg2] .. [argN]
```
Need to check the formatting here.

rlubos commented on drivers/wifi/nrf700x/zephyr/src/qspi/inc/ficr_prog.h:
```diff
@@ -0,0 +1,86 @@
+#define     CALIB_MAXPOW5G0MCS0     81
+#define     CALIB_RXGAINOFFSET      82
+#define     CALIB_TXPOWBACKOFFT     83
+#define     CALIB_TXPOWBACKOFFV     84
+#define     REGION_DEFAULTS			85
+#define     OTP_MAX_WORD_LEN		128
```
Those alignments look messy, please fix them (same applies to other defines below). I suspect it could be editor misconfiguration on your side? We use 8 spaces/tab in NCS.

rlubos commented on drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c:
```diff
@@ -0,0 +1,426 @@
+		(buff[off3] == 0xFFFFFFFF) &&
+		(buff[off4] == 0xFFFFFFFF))
+		return 0xFFFFFFFF;
+	else if ((buff[off1] == 0x50FA50FA) &&
```
Could we define symbols for those values instead of using magic numbers? I saw they're also used in the shell part.

bama-nordic replied:
Done

rlubos commented on drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c:
```diff
@@ -0,0 +1,426 @@
+
+static int req_otp_read_mode(void)
+{
+
```
Unneeded empty line.

bama-nordic replied:
Done

rlubos commented on drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c:
```diff
@@ -0,0 +1,426 @@
+
+static int req_otp_byte_write_mode(void)
+{
+
```
Unneeded empty line.

bama-nordic replied:
Done

rlubos commented on drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c:
```diff
@@ -0,0 +1,426 @@
+
+static unsigned int read_otp_location_disp(unsigned int offset, unsigned int *read_val)
+{
+
```
Unneeded empty line.

rlubos commented on drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c:
```diff
@@ -0,0 +1,426 @@
+int write_otp_memory(unsigned int otp_addr, unsigned int *write_val)
+{
+	int err;
+	int	mask_val;
```
There's some weird offset here, please use space instead of tab.

rlubos commented on samples/wifi/radio_test/src/nrf_wifi_radio_ficr_shell.c:
```diff
@@ -0,0 +1,345 @@
+
+static void disp_location_status(char *region, unsigned int ret)
+{
+
```
Unneeded empty line.

rlubos commented on samples/wifi/radio_test/src/nrf_wifi_radio_ficr_shell.c:
```diff
@@ -0,0 +1,345 @@
+{
+
+	switch (ret) {
+
```
Unneeded empty line.

rlubos commented on samples/wifi/radio_test/src/nrf_wifi_radio_ficr_shell.c:
```diff
@@ -0,0 +1,345 @@
+		break;
+	}
+
+	LOG_INF("\n");
```
Is this empty line really needed?

bama-nordic replied:
Yes, to make output a bit readable

rlubos commented on samples/wifi/radio_test/src/nrf_wifi_radio_ficr_shell.c:
```diff
@@ -0,0 +1,345 @@
+static void disp_fields_status(unsigned int flags)
+{
+	if (flags & (~QSPI_KEY_FLAG_MASK))
+		LOG_INF("QSPI Keys are not programmed in OTP\n");
```
Please use surrounding `{ }` even for single-line blocks.

bama-nordic replied:
This was there originally but I think checkpatch gave me warning so I changed it!! 
But sure, will change it back.

rlubos replied:
For Zephyr (and thus NCS), you need to run checkpatch with `--ignore BRACES`, see https://docs.zephyrproject.org/3.1.0/contribute/guidelines.html#coding-style (first bullet).

bama-nordic replied:
Done

rlubos commented on samples/wifi/radio_test/src/nrf_wifi_radio_ficr_shell.c:
```diff
@@ -0,0 +1,345 @@
+{
+	unsigned int val[OTP_MAX_WORD_LEN];
+	unsigned int ret, err;
+
```
Unneeded empty line.

rlubos commented on samples/wifi/radio_test/src/nrf_wifi_radio_ficr_shell.c:
```diff
@@ -0,0 +1,345 @@
+		return -ENOEXEC;
+	}
+
+	ret = check_protection(&val[0], REGION_PROTECT+0, REGION_PROTECT+1,
```
Please add spaces around operators (`REGION_PROTECT + 0` etc).

rlubos commented on samples/wifi/radio_test/src/nrf_wifi_radio_ficr_shell.c:
```diff
@@ -0,0 +1,345 @@
+	LOG_INF("\n");
+
+	LOG_INF("CALIB_XO = 0x%02x\n", val[CALIB_XO]&0xff);
+	LOG_INF("\n");
```
Again, are those empty lines in between really needed?

bama-nordic replied:
Yes, readability

rlubos commented on samples/wifi/radio_test/src/nrf_wifi_radio_ficr_shell.c:
```diff
@@ -0,0 +1,345 @@
+
+	disp_location_status("OTP", ret);
+
+	LOG_INF("REGION_PROTECT0 = 0x%08x\n", val[REGION_PROTECT+0]);
```
A general comment regarding logger usage in this module. It seems that it is used to print repsponse for the shell command. It could be the case, that logger and shell are configured to use different mediums (for example shell over UART / logger over RTT). In such case user won't see the command output.

A better choice here would be to use `shell_fprintf()` function, to make sure the command output is printed over the same medium as shell.

nvlsianpu commented on drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c:
```diff
@@ -0,0 +1,426 @@
+		(buff[off4] == 0x50FA50FA))
+		return 0x50FA50FA;
+	else
+		return 0xdeadbeef;
```
Nice joke :D
Unfortunately we don't have any sense of humor.
Pleas us a sad predefined value here

nvlsianpu replied:
For all of returned values here.

nvlsianpu commented on drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c:
```diff
@@ -0,0 +1,426 @@
+	int otp_mem_status = 0;
+	int poll = 0;
+
+	while (poll != 100) {
```
I would use a timeout instead of counting toward 100

bama-nordic replied:
This is a tested code that needs to go for customer sampling (needs to get merged today). Look into this for then next revision

nvlsianpu commented on drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c:
```diff
@@ -0,0 +1,426 @@
+	while (poll != 100) {
+		read_word(OTP_POLL_ADDR, &otp_mem_status);
+
+		if ((otp_mem_status&0x4) == OTP_READY) {
```
```suggestion
		if ((otp_mem_status & OTP_READY) == OTP_READY) {
```
or better introduce #define OTP_READY_MASK 0x4

nvlsianpu commented on drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c:
```diff
@@ -0,0 +1,426 @@
+
+static int req_otp_standby_mode(void)
+{
+	write_word(OTP_RWSBMODE_ADDR, 0x0);
```
use OTP_STANDBY_MODE

bama-nordic replied:
Nope, this is used for READ, WRITE and STANDBY mode

nvlsianpu commented on drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c:
```diff
@@ -0,0 +1,426 @@
+	int otp_mem_status = 0;
+	int poll = 0;
+
+	while (poll != 100) {
```
for (

bama-nordic replied:
This is a tested code that needs to go for customer sampling (needs to get merged today). Look into this for then next revision

nvlsianpu commented on drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c:
```diff
@@ -0,0 +1,426 @@
+	return -ENOEXEC;
+}
+
+static int poll_otp_wrdone(void)
```
poll_otp_wrdone(), poll_otp_read_valid(), poll_otp_ready() are very similar. Their code can be commonized.

bama-nordic replied:
This is a tested code that needs to go for customer sampling (needs to get merged today). Look into this for then next revision

nvlsianpu commented on drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c:
```diff
@@ -0,0 +1,426 @@
+	while (poll < 100) {
+		read_word(OTP_POLL_ADDR, &otp_mem_status);
+
+		if ((otp_mem_status&0x1) == OTP_WR_DONE) {
```
& `mask`

nvlsianpu commented on drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c:
```diff
@@ -0,0 +1,426 @@
+	while (poll < 100) {
+		read_word(OTP_POLL_ADDR, &otp_mem_status);
+
+		if ((otp_mem_status&0x2) == OTP_READ_VALID) {
```
& `mask`

nvlsianpu commented on drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c:
```diff
@@ -0,0 +1,426 @@
+	return poll_otp_ready();
+}
+
+static unsigned int read_otp_location_disp(unsigned int offset, unsigned int *read_val)
```
for what entity `disp` standing here?

bama-nordic replied:
Changed to read_otp_location()

nvlsianpu commented on drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c:
```diff
@@ -0,0 +1,426 @@
+	switch (otp_addr) {
+
+	case REGION_PROTECT:
+		write_otp_location(REGION_PROTECT, write_val[0]);
```
this case block can be done in a loop/s

nvlsianpu commented on drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c:
```diff
@@ -0,0 +1,426 @@
+		break;
+
+	case QSPI_KEY:
+		mask_val = QSPI_KEY_FLAG_MASK;
```
this case block can be done in a loop/s

bama-nordic replied:
This is a tested code that needs to go for customer sampling (needs to get merged today). Look into this for then next revision

rlubos commented:
@bama-nordic Something must've gone wrong with the recent push, there are plenty of compliance errors.

richabp commented on doc/nrf/releases/release-notes-changelog.rst:
```diff
@@ -240,6 +240,7 @@ Wi-Fi samples
 -------------
 
 * Added :ref:`wifi_radio_test` sample with the radio test support.
+    - Included subcommand for FICR/OTP programming
```
```suggestion

  * Added subcommands for FICR/OTP programming.
```

bama-nordic replied:
OK, changed.

richabp reviewed (changes requested):
Minor change.

bama-nordic commented:
> @bama-nordic Something must've gone wrong with the recent push, there are plenty of compliance errors.

Thanks - Yes, some spaces when I updated output logs in the documentation. Fixed now - should hopefully be fine.
